File: examples/scala-zio/Welcome.scala
// Welcome effect example in Scala with ZIO.
// Two effects: Ask and Tell.
// The implementations (layers) in run are swappable without changing the logic.

import zio._
import zio.Console.{printLine, print => printNoNewline, readLine}


// --- Service interfaces ---

trait Ask:
  def ask(prompt: String): UIO[String]

trait Tell:
  def tell(message: String): UIO[Unit]


// --- Accessor methods (lift each service method into a ZIO) ---

object Ask:
  def ask(prompt: String): ZIO[Ask, Nothing, String] =
    ZIO.serviceWithZIO[Ask](_.ask(prompt))

object Tell:
  def tell(message: String): ZIO[Tell, Nothing, Unit] =
    ZIO.serviceWithZIO[Tell](_.tell(message))


// --- Core logic (description — nothing runs yet) ---

val greet: ZIO[Ask & Tell, Nothing, Unit] =
  for
    name <- Ask.ask("What is your name? ")
    _    <- Tell.tell(s"Hello, $name!")
  yield ()


// --- Implementations ---

val consoleAsk: ULayer[Ask] = ZLayer.succeed(new Ask:
  def ask(prompt: String): UIO[String] =
    printNoNewline(prompt).orDie *> readLine.orDie)

val consoleTell: ULayer[Tell] = ZLayer.succeed(new Tell:
  def tell(message: String): UIO[Unit] = printLine(message).orDie)


// --- Entry point ---

object Main extends ZIOAppDefault:
  def run = greet.provide(consoleAsk, consoleTell)


// =============================================================================
// WHAT'S NEW IN THIS EXAMPLE
// =============================================================================
//
// FOR-COMPREHENSION (SEQUENCING EFFECTS)
//
//   val greet: ZIO[Ask & Tell, Nothing, Unit] =
//     for
//       name <- Ask.ask("What is your name? ")
//       _    <- Tell.tell(s"Hello, $name!")
//     yield ()
//
// ZIO's `for`/`yield` comprehension sequences effects. The `<-` arrow binds
// the result of each ZIO to a variable. `name` receives the String produced
// by `Ask.ask`. The underscore `_` discards the Unit result of `Tell.tell`.
// `yield ()` ends the comprehension with a Unit value. Each step runs in
// order; later steps can reference earlier bindings.
//
//
// ENVIRONMENT INTERSECTION
//
//   ZIO[Ask & Tell, Nothing, Unit]
//
// `Ask & Tell` is a Scala type intersection: `greet` requires both services
// to be present in its environment.
//
//
// PROVIDING MULTIPLE LAYERS
//
//   greet.provide(consoleAsk, consoleTell)
//
// `.provide(...)` accepts multiple layers. ZIO combines them and wires each
// service into the environment, satisfying the `Ask & Tell` requirement.
//
//
// SEQUENCING WITHIN AN IMPLEMENTATION
//
//   printNoNewline(prompt).orDie *> readLine.orDie
//
// `*>` is ZIO's sequence-and-discard operator: run the left effect, discard
// its result, then run the right effect and return its result. Here: print
// the prompt (no result needed), then read a line and return the string.
//
//
// STRING INTERPOLATION
//
//   s"Hello, $name!"
//
// The `s` prefix enables string interpolation; `$name` is replaced by the
// variable's value.

File: examples/effect-ts/welcome.ts
// Welcome effect example in TypeScript with Effect.ts.
// Two effects: Ask and Tell.
// The implementations (layers) in runPromise are swappable without changing the logic.

import { Context, Effect, Layer } from "effect"
import * as readline from "node:readline"
import * as process from "node:process"


// --- Service tags ---

class Ask extends Context.Tag("Ask")<
  Ask,
  { ask: (prompt: string) => Effect.Effect<string> }
>() {}

class Tell extends Context.Tag("Tell")<
  Tell,
  { tell: (message: string) => Effect.Effect<void> }
>() {}


// --- Core logic ---

const greet = Effect.gen(function* () {
  const ask  = yield* Ask
  const tell = yield* Tell
  const name = yield* ask.ask("What is your name? ")
  yield* tell.tell(`Hello, ${name}!`)
})


// --- Implementations ---

const ConsoleAsk = Layer.succeed(Ask, {
  ask: (prompt) =>
    Effect.async<string>((resolve) => {
      const rl = readline.createInterface({
        input:  process.stdin,
        output: process.stdout,
      })
      rl.question(prompt, (answer) => {
        rl.close()
        resolve(Effect.succeed(answer))
      })
    }),
})

const ConsoleTell = Layer.succeed(Tell, {
  tell: (message) => Effect.sync(() => console.log(message)),
})


// --- Entry point ---

Effect.runPromise(
  greet.pipe(Effect.provide(Layer.merge(ConsoleAsk, ConsoleTell))),
)


// =============================================================================
// WHAT'S NEW IN THIS EXAMPLE
// =============================================================================
//
// RESOLVING MULTIPLE SERVICES
//
//   const ask  = yield* Ask
//   const tell = yield* Tell
//
// Each `yield* Tag` resolves one service from the environment. With two
// services, both are resolved at the start of the generator before they
// are used. The order of resolution does not matter.
//
//
// COMBINING LAYERS
//
//   greet.pipe(Effect.provide(Layer.merge(ConsoleAsk, ConsoleTell)))
//
// `Layer.merge` combines two layers so both services are available when the
// effect runs.
//
//
// WRAPPING CALLBACK-BASED ASYNC
//
//   Effect.async<string>((resolve) => {
//     const rl = readline.createInterface({ ... })
//     rl.question(prompt, (answer) => {
//       rl.close()
//       resolve(Effect.succeed(answer))
//     })
//   })
//
// `Effect.async` bridges callback-based APIs into the Effect world. The
// outer function receives a `resolve` callback. When the async operation
// completes (here: when the user presses Enter), call
// `resolve(Effect.succeed(value))` to deliver the result back into the
// Effect runtime. Until `resolve` is called, the Effect is suspended.
